cmd/ocsp-updater/main.go

```diff
@@ -51,28 +102,70 @@ func setupClients(c cmd.Config, stats statsd.Statter) (rpc.CertificateAuthorityC
 	return cac, closeChan
 }
 
-func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string) error {
-	certObj, err := tx.Get(core.Certificate{}, serial)
-	if err != nil {
-		return err
-	}
-	statusObj, err := tx.Get(core.CertificateStatus{}, serial)
-	if err != nil {
-		return err
+var errNoStaleResponses = errors.New("No stale responses to update")
+
+func (updater *OCSPUpdater) findStaleOCSPResponses(oldestLastUpdatedTime time.Time, batchSize int) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT cert.*
+		 FROM certificateStatus AS cs
+		 JOIN certificates AS cert
+		 ON cs.serial = cert.serial
+		 WHERE cs.ocspLastUpdated < :lastUpdate
+		 AND cert.expires > now()
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT :limit`,
+		map[string]interface{}{
+			"lastUpdate": oldestLastUpdatedTime,
+			"limit":      batchSize,
+		},
+	)
+	if err == sql.ErrNoRows || len(certs) == 0 {
+		return certs, errNoStaleResponses
 	}
+	return certs, err
+}
 
-	cert, ok := certObj.(*core.Certificate)
-	if !ok {
-		return fmt.Errorf("Cast failure")
+var errNoNewCertificates = errors.New("No certificates issued since last check")
+
+func (updater *OCSPUpdater) getCertificatesIssuedSince(since time.Time, batchSize int) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT * FROM certificates
+		 WHERE issued > :since
```

Comment:
Actually, I take that back. This is quite hard because we wind up introducing a dependency on highly accurate time between two different machines. We should plan to function correctly even in a world where times on the CA and OCSP Updater machines are several seconds apart.

The problem with what I proposed above is that the OCSP Updater may run a tick at 12:09:31 pm, and asked for all certs issued before 12:09:31. On the next tick it might ask for all certs issued after 12:09:31 and before 12:09:32. However, if the CA is a few seconds behind, it might issue a cert with `notBefore` of 12:09:30 in the meantime, and that cert would be missed.